src/plugins/janus_sip.c

```diff
@@ -7392,6 +7401,60 @@ gpointer janus_sip_sofia_thread(gpointer user_data) {
 	return NULL;
 }
 
+/* Check peer RTP has RFC2833 and push event */
+static void janus_sip_check_rfc2833(janus_sip_session *session, char *buffer, int len) {
+	janus_rtp_header *rtp_header = (janus_rtp_header *)buffer;
+	if (rtp_header->type != JANUS_RTP_RFC2833_DEFAULT_PAYLOAD_TYPE)
+		return;
+	int plen = 0;
+	char *payload_buffer = janus_rtp_payload(buffer, len, &plen);
+	int rfc2833_payload_size = sizeof(janus_rtp_rfc2833_payload);
+	if(plen < rfc2833_payload_size)
+		return;
+	janus_rtp_rfc2833_payload *rfc2833_payload = (janus_rtp_rfc2833_payload *)payload_buffer;
+	uint16_t duration = ntohs(rfc2833_payload->duration);
+	if (rfc2833_payload->end == 0)
+		return;
+
+	/* Set up last dtmf to avoid duplication */
+	if(session->latest_dtmf != NULL && session->latest_dtmf->dtmf_event_id == rfc2833_payload->event && session->latest_dtmf->timestamp == rtp_header->timestamp)
+		return;
+	if(session->latest_dtmf != NULL){
+		g_free(session->latest_dtmf);
+		session->latest_dtmf = NULL;
+	}
+	session->latest_dtmf = g_malloc0(sizeof(janus_sip_dtmf));
+	session->latest_dtmf->dtmf_event_id = rfc2833_payload->event;
+	session->latest_dtmf->timestamp = rtp_header->timestamp;
+
+	/* Parse dtmf key */ 
+	uint16_t dtmf_key;
+	uint16_t dtmf_keys[] = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '*', '#', 'A', 'B', 'C', 'D'};
```

Comment:
I guess this could be a static array defined as part of the plugin globals: no need to redefine it any time we invoke this function, those values never change.